File: examples/cpp/pd0FileDemonstrator/src/pd0filedemonstrator.cpp
#include <stdlib.h>
#include <iostream>

// Core PDN message library
#include "PDN_messages.h"

// Custom user PD0 section
#include "UserPD0Block.h"

// Utility function
long find_file_size(const char* filename);

using namespace sonardyne_edge;

using namespace PDN_messages;

int main(int argc, char** argv) 
{
    
    if (argc < 2)
    {
        std::cout << "Usage: ./pd0filedemonstrator <filename> [Optional: output.filename] " << std::endl;
        std::cout << "       ./pd0filedemonstrator <filename> " << std::endl;
        std::cout << "       ./pd0filedemonstrator <input filename> <output filename>" << std::endl;
        
        return EXIT_FAILURE;
    }
    
    std::string filename(argv[1]);
    std::cout << "Decoding PD0 file " << filename << std::endl;

    if (argc == 2)
    {
        std::cout << "Adding user block to deserialise pd0" << std::endl;        
    }

    // In this simple demo, if an output file is specified, an example custom user PD0 section 
    // will be inserted (see UserPD0Block.h/cpp for the declaration/definition)
    FILE* fh_out = NULL;
    if (argc > 2)
    {
        std::cout << "Adding user custom block to output pd0" << std::endl;
        fh_out = fopen(argv[2], "w");    
    }
    
    long file_size_bytes = find_file_size(filename.c_str());
       
    // first, peek the header to see how much data to read
    if (file_size_bytes >= RDIPD0Header::SIZE_BYTES)
    {
        // open the file
        FILE* fh = fopen(filename.c_str(), "r");

        if (fh)
        {
            int file_bytes_read = 0;

            while (file_bytes_read < file_size_bytes)
            {   
                // first, read the header to figure out how much data there is
                char hdr_buf[RDIPD0Header::SIZE_BYTES];
                RDIPD0Header hdr;

                fread(&hdr_buf[0], 1, RDIPD0Header::SIZE_BYTES, fh);

                hdr.deserialise(&hdr_buf[0], RDIPD0Header::SIZE_BYTES);

                // seek the file back to the start of the header
                fseek(fh, -RDIPD0Header::SIZE_BYTES, SEEK_CUR);

                // read in the ensemble
                char ensemble_buf[hdr.EnsembleBytes + PD0Message::CHECKSUM_SIZE_BYTES];

                int bytes_read = fread(&ensemble_buf[0], 1, hdr.EnsembleBytes + PD0Message::CHECKSUM_SIZE_BYTES, fh);                    

                if (bytes_read > 0)
                {
                    std::cout << "Read " << bytes_read << " bytes ; total read: " << file_bytes_read << " / " << file_size_bytes << std::endl;

                    // We have the whole ensemble in memory. Deserialise it.
                    PD0Message pd0;     

                    // if a known user section is present in the PD0, add it here before deserialisation
                    if (argc == 2)
                    {
                        UserPD0Block* tmp_user = new UserPD0Block();
                        pd0.add_section(tmp_user);
                    }

                    int bytes_deserialised = pd0.deserialise(&ensemble_buf[0], hdr.EnsembleBytes + PD0Message::CHECKSUM_SIZE_BYTES);

                    SONOriginSystemLeader*       pSONVar        = (SONOriginSystemLeader*)pd0.get_section(SONOriginSystemLeader::Id);
                    if(pSONVar != NULL)
                    {
                        std::cout << "ping counter = " << pSONVar->u32_PingCounter << std::endl;
                    }
                    if (fh_out)
                    {
                        // if an output file is specified, add a custom user PD0 section
                        if (argc == 3)
                        {
                            UserPD0Block* user = new UserPD0Block();
                            pd0.add_section(user);                        
                        }

                        EFrame eFrame = EFrame::EARTH;
                        pd0.transform(eFrame);

                        int new_length_bytes = pd0.calc_length();

                        char* buf = new char[new_length_bytes];

                        int bytes_serialised = pd0.serialise(buf, new_length_bytes);

                        fwrite(buf, 1, bytes_serialised, fh_out);

                        delete[] buf;
                    }


                    file_bytes_read += bytes_read;
                }
                else
                {
                    std::cout << "Read " << bytes_read << " bytes ; total read: " << file_bytes_read << " / " << file_size_bytes << std::endl;
                    std::cout << "File read error: " << bytes_read << std::endl;
                    break;
                }
            } 

            std::cout << "Finished: Read " << file_bytes_read << " / " << file_size_bytes << " bytes" << std::endl;

            fclose(fh);
        }
    }
    else
    {
        std::cout << "File too small" << std::endl;
        return -1;
    }

    return EXIT_SUCCESS;
}

long find_file_size(const char* filename)
{
    long bytes = -1;
    
    FILE* fh = fopen(filename, "r");
    
    if (fh)
    {
        int error = fseek(fh, 0, SEEK_END);
    
        if (error == 0)
        {
            bytes = ftell(fh);
        }            
        
        fclose(fh);
    }
    
    return bytes;
}

File: examples/python3/pd0FileDemonstrator.py
import sys
import os
import math

sys.path.insert(1, '../../src/')

from PDNMessages.PDN_messages   import *
from PDNMessages.UserPD0Section import *

if len(sys.argv) == 2:

    path = sys.argv[1]

    filesize = os.path.getsize(path)

    with open(path, "rb", buffering = 0) as f:    

        # open an output file for writing
        output_file = open(path + ".out", 'wb')

        # figure out how big the ensemble is from the 6-byte header
        hdr = RDIPD0Header()
        cnt = 0
        num_ensembles = 0
        
        while (data := f.read(RDIPD0Header.SIZE_BYTES)):

            print("===== New ensemble =====")

            pd0 = PD0Message()

            # ---------------- read header ----------------- #
            # put data into byte array
            buf = bytearray(data)
            
            bytes_deserialised = pd0.deserialise(buf)

            ensemble_bytes = pd0.get_section(RDIPD0Header.ID).EnsembleBytes + PD0Message.CHECKSUM_SIZE_BYTES
            print("Ensemble size = "+ str(ensemble_bytes))
            if (cnt == 0):
                num_ensembles = math.ceil(filesize / ensemble_bytes)

            # seek back on the header size
            f.seek(-RDIPD0Header.SIZE_BYTES, 1)


            # ---------------- read entire ensemble ----------------- #
            data = f.read(ensemble_bytes)

            # put data into byte array
            orig_ensemble_in_buf = bytearray(data)
            
            bytes_deserialised = pd0.deserialise(orig_ensemble_in_buf)

            pd0.show()  # will print each Pd0 content.
            
            print(str(bytes_deserialised) + " bytes de-serialised")


            # ---------------- re-serialise the ensemble ----------------- #
            
            orig_ensemble_out_buf = bytearray(pd0.calc_length())

            bytes_serialised = pd0.serialise(orig_ensemble_out_buf)
            
            print(str(bytes_serialised) + " bytes re-serialised")


            # ---------------- add user-defined PD0 section --------- #
            
            # add a user field to modify the output PD0
            user = UserPD0Section()
            user.UserInt    = 1;
            user.UserUInt   = 2;
            user.UserInt64  = 3;
            user.UserFloat  = 4.56;
            user.UserDouble = 7.89;

            pd0.add_section(user)


            # ---------------- serialise the modified ensemble ----------------- #
            
            modified_ensemble_out_buf = bytearray(pd0.calc_length())

            bytes_serialised = pd0.serialise(modified_ensemble_out_buf)
            
            print(str(bytes_serialised) + " bytes serialised of modified ensemble")


            # ----- deserialise the modified ensemble to show how it's done ---- #
            
            # we add the user section so PD0 knows how to decode it
            new_pd0 = PD0Message()
            user = UserPD0Section() 
            new_pd0.add_section(user)

            bytes_deserialised = new_pd0.deserialise(modified_ensemble_out_buf)

            print(str(bytes_deserialised) + " bytes de-serialised of modified ensemble")

            # --------------- write the modified ensemble to disk -------------- #

            output_file.write(modified_ensemble_out_buf)

            cnt = cnt + 1
            
            print("Processing ensemble " + str(cnt) + " / " + str(num_ensembles))

    # close the output file
    output_file.close()

else:

    print("Usage: ")
    print("python3 pd0demonstrator.py /path/to/file.pd0")
